### DISCUSSION TITLE
Self-Relation, Self-Referencing Type

### DISCUSSION THREAD

**Discussion by v4nderstruck** [prisma/prisma#18548] [Q&A] [answered]
Is there a way to build Self-Referencing Types for Self-Relations using Prisma? 

Let's say I have a model **One-To-Many Self Relation** like this

```prisma
model User {
  id        Int     @id @default(autoincrement())
  name      String?
  teacherId Int?
  teacher   User?   @relation("TeacherStudents", fields: [teacherId], references: [id])
  students  User[]  @relation("TeacherStudents")
}
```

Using `type UserWithStudents = Prisma.UserGetPayload<{include: {students: true}}>` we would end up with a type like this

```typescript
type UserWithStudents = User & {
   students: User[];
  // scalar fields...
}
```

However, it would be convenient to have a **Self-Referencing** Type because of the Self-Relation... I have not found anything in the docs.

```typescript
type UserWithStudents = User & {
   students: UserWithStudents[];
  // scalar fields...
}
```

**Comment by nurul3101** [comment]
Hello @zensayyy 👋 

This is indeed an interesting question, to make sure that I understand your question correctly,

For now, the generated type would look like this:
<img width="646" alt="Screenshot 2023-03-29 at 5 49 29 PM" src="https://user-images.githubusercontent.com/13049676/228533286-02c42614-9f4a-4a06-b4f4-85dbf14377dd.png">

But you expect the generated type to look something like this, correct?
<img width="613" alt="Screenshot 2023-03-29 at 5 50 41 PM" src="https://user-images.githubusercontent.com/13049676/228533569-34c0be10-dffc-40bb-a3b5-59f1b0d3518b.png">

**Reply by v4nderstruck** [reply]
Yes exactly. Currently, this does not allow you to go deeper than one student level. It would be interesting to know if there is a way to get the nice self referencing type from the Prisma generated types. We may have to create a feature request if that is not possible yet!

**Comment by v4nderstruck** [comment, marked as answer]
Well, I found out the intended way to do this. You should do 

```typescript
type UserWithStudents = Prisma.UserGetPayload<{
   include: {
       students: { 
            include: {students: true}  
       }
   }
}>

// which gives you something like
type UserWithStudents = User & {
    students: ( User & {students: User[] } )[]
}

```

Hence, I will close this one.
